### DISCUSSION TITLE
zfsbootmenu loop

### DISCUSSION THREAD

**Discussion by bobblkabb** [zbm-dev/zfsbootmenu#296] [General]
I have installed according to the instructions [Void Linux Single disk syslinux MBR](https://github.com/zbm-dev/zfsbootmenu/wiki/Void-Linux----Single-disk-syslinux-MBR) (encrypted), but I end up in a loop in the zfsbootmenu. 
When I press return the timer starts again at 5s.
What could have gone wrong here?

**Comment by Dragon-debian** [comment]
does it import any pool?
whats wrote on screen?
press `Esc` and execute `zpool list`. does it show any imported pool?

**Comment by zdykstra** [comment]
There's only one 5 second loop in ZFSBootMenu, and that's when it can't import any pools. There's an option to drop to a recovery shell via the `escape` key. Once you're in the recovery shell please run `zpool import` and `zpool list` and relay exactly what the commands show.

**Comment by bobblkabb** [comment]
> does it import any pool? whats wrote on screen? press `Esc` and execute `zpool list`. does it show any imported pool?

> There's only one 5 second loop in ZFSBootMenu, and that's when it can't import any pools. There's an option to drop to a recovery shell via the `escape` key. Once you're in the recovery shell please run `zpool import` and `zpool list` and relay exactly what the commands show.


It says "Loading zpool ... Failed, no file or directory".
When should the encryption kick in btw? Before or after zfsbootmenu?..

**Comment by zdykstra** [comment]
Please reboot your system and tell me _exactly_ what countdown screen shows. We can go from there once you've confirmed the basics.

**Comment by bobblkabb** [comment]
> Please reboot your system and tell me _exactly_ what countdown screen shows. We can go from there once you've confirmed the basics.

I see the "Boot Menu".
There are a few entries in the meantime because I tried to fix it. So now I have
```
ZFSBootMenu 1.12.0_5
ZFSBootMenu 1.12.0_4
ZFSBootMenu 1.12.0_3
ZFSBootMenu 1.12.0_2
```

And on the bottom there's a timer counting from 5 to 0 over and over.

**Comment by zdykstra** [comment]
Okay, you're stuck in syslinux and not ZFSBootMenu. That's a completely different problem that is very difficult to debug. Does your system not support UEFI ?

**Comment by bobblkabb** [comment]
The laptop is a little older and I couldn't get UEFI work on it. System wasn't recognized, it always jumped to the next available boot option. Installed it a few times.
I installed the pre-build zfsbootmenu binary. refind-install gave me an error and suggested to rename refind_x64.efi to bootx64.efi, which I did. It also showed some efi variable errors. efivar-tester said it doesn't support efi variables or so. 
Fiddled with Bios settings (CSM -> enabled/disabled) but without success.

**Comment by zdykstra** [comment]
Unfortunately, syslinux can be very hit-or-miss.  The only suggest I have would be to do the following:

* Rename `/boot/syslinux/zfsbootmenu` to `/boot/syslinux/zbm` on your syslinux partition.
* Edit `/etc/zfsbootmenu/config.yaml` and change `ImageDir` to `/boot/syslinux/zbm`.
* Re-run `generate-zbm` and after that, verify that `/boot/syslinux/syslinux.conf` has the corrected paths in the file.

If that doesn't work, you'll have to dig around external syslinux documentation to see if you can find any clues as to what might be going wrong.

**Comment by bobblkabb** [comment]
There is only /boot/syslinux - nothing inside! So here's the problem!

Or did I do anything wrong to access the system?:

1. Boot hrmpf system from USB
2. xbps-reconfigure -a
modprobe zfs
3. zpool import -N -R /mnt zroot
zfs load-key -L prompt zroot
zfs mount zroot/ROOT/void
zfs mount zroot/home
4. mount -t proc proc /mnt/proc
mount -t sysfs sys /mnt/sys
mount -B /dev /mnt/dev
mount -t devpts pts /mnt/dev/pts
chroot /mnt /bin/bash

**Comment by zdykstra** [comment]
You wouldn't see a syslinux menu with ZFSBootMenu items in it if that were the case. You just need to mount the partition/filesystem that holds /boot/syslinux. Probably /dev/sda1.

**Comment by bobblkabb** [comment]
It's on sda1 - of course! 
Now I renamed, edited and checked but still the same behavior. 

Maybe I should try to install it on different hardware.

**Comment by Dragon-debian** [comment]
@bobblkabb
can you put the `config.yaml` content here?

**Comment by bobblkabb** [comment]
config.yaml:
```
Global:
  ManageImages: true
  BootMountPoint: /boot/syslinux
  DracutConfDir: /etc/zfsbootmenu/dracut.conf.d
  PreHooksDir: /etc/zfsbootmenu/generate-zbm.pre.d
  PostHooksDir: /etc/zfsbootmenu/generate-zbm.post.d
Components:
  ImageDir: /boot/syslinux/zbm
  Versions: 3
  Enabled: true
  syslinux:
    Config: /boot/syslinux/syslinux.cfg
    Enabled: true
EFI:
  ImageDir: /boot/efi/EFI/void
  Versions: false
  Enabled: false
Kernel:
  CommandLine: ro quiet loglevel=0
```

**Comment by Dragon-debian** [comment]
your `config.yaml` seems ok.


> It's on sda1 - of course! Now I renamed, edited and checked but still the same behavior.
> 
> Maybe I should try to install it on different hardware.

`generate-zbm` unmounts `$BootMountPoint` after its job is done, if it wasnt mounted before running `generate-zbm`.

mount `/boot/syslinux` somewhere (e.g. /mnt) and list its content. according to your `config.yaml`, there should be at least a `syslinux.cfg` in there, plus `zbm` directory which holds kernel and initramfs of ZFSBootMenu.

check `syslinux.cfg` content and see if it matches the available kernel and initramfs pairs.

**Comment by bobblkabb** [comment]
```
mount /dev/sda1 /mnt/sda1
ls -l /mnt/sda1
```
--> syslinux.cfg is in it and there's also that subfolder zbm:
<a href="https://imgur.com/fwCN036"><img src="https://i.imgur.com/fwCN036t.jpg" title="source: imgur.com" /></a>

**Comment by ahesford** [comment]
Try changing "KERNEL" to "LINUX" in your `syslinux.cfg`.

**Comment by bobblkabb** [comment]
Didn't work, still the same. Or do I need to execute a command after I changed it?

**Comment by Dragon-debian** [comment]
Try to load the kernel and initramfs manually.
In syslinux menu, press `Escape`, you should see a `boot: ` prompt. type this, it should load the kernel and initramfs. see what happens.
```
.linux /zbm/vmlinuz-1.12.0_13 initrd=/zbm/initramfs-1.12.0_13.img
```
you may need to replace `1.12.0_13` with something else.
use `ls /zbm` in syslinux to find the right name

**Comment by bobblkabb** [comment]
Loading manually doesn't work - 'no directory / file found..'.
'ls' show syslinux.cfg and many other files. There's also the zbm folder but when I do 'ls /zbm' or 'cd zbm' it fails, 'no such file or directory' (?!)
When I boot the system via hrmpf I can access this folder.

**Comment by ahesford** [comment]
I have seen this issue before and am not aware of a solution. Things you can try:
- Move the ZBM kernel and initramfs out of `zbm` and into the `syslinux` root
- Try replacing the ext4 Stalinist partition with another filesystem that suslinux can read; maybe ext3 or ext2 will give you more luck.

In any event, this is definitely a syslinux problem. If you want to keep discussing this, please take this up in the "Discussions" page above, or join us on IRC at `#zfsbootmenu` on libera.chat.

**Comment by bobblkabb** [comment]
I'm getting a bit closer to it. In fact, it doesn't work with ext4. I tried a second time and then chose ext3. 

However, I now encounter the next problem: after the zfs boot menu, I get to a screen where it says in the middle that you can boot with Return and get to the boot menu with ESC. Here, the system no longer accepts keyboard input. After the timeout, the password is requested (_Enter passphrase for 'zroot':_). Here, too, no input is accepted.

**Comment by Dragon-debian** [comment]
try fat or vfat (fat32) for syslinux partition. Im using ZFSBootMenu + syslinux on an old machine (from 2007). the syslinux partition is formated with vfat.
also enable `usb keyboard` support in bios if there is any. I wasnt able to use keyboard in syslinux until I enabled that.

**Comment by bobblkabb** [comment]
Formated sda1 to vfat but still the same.
Toggled those Bios settings:

```
Legacy USB Support - Enabled/Disabled
USB Port - Enabled/Disabled
XHCI Controller Setting - Standard/Compatible
Under 'Intel Management Engine Configuration':
USB Provisioning - Enabled/Disabled
```

**Comment by bobblkabb** [comment]
I'll try another UEFI installation. Made progress on a different laptop.

**Comment by bobblkabb** [comment]
Again, laptop ignores keyboard input.

**Comment by bobblkabb** [comment]
Bios Update didn't help. Was nearly 5 years old, over 10 iterations.

**Comment by ahesford** [comment]
These are not ZBM problems, but a dracut problem. You are either missing keyboard support in the ZBM dracut image or the keyboard is not properly configured. On some laptops, to get the keyboard working in the initramfs, I need to play around with some of the `i8042` module options. On a modern laptop, I use
```
i8042.dumbkbd i8042.noaux i8042.nomux
```
There are many other options; `modinfo i8042` will show you descriptions.

**Comment by bobblkabb** [comment]
Great. That's it. Thanks! Keyboard works. 
Now I'm getting the same error like described [here](https://github.com/zbm-dev/zfsbootmenu/discussions/297) on a second laptop.
